matchzoo/preprocessors.py

```diff
@@ -1,146 +1,165 @@
 """Matchzoo toolkit for text pre-processing."""
 
 import re
+import abc
 import nltk
-import jieba
 import typing
 import many_stop_words
-from functools import reduce
 
 
 match_punc = re.compile('[^\w\s]')
 
 
-def get_stopwords(lang: str='en') -> list:
-    """
-    Get stopwords based on language.
+class StatelessProcessUnit(metaclass=abc.ABCMeta):
+    """Process unit do not persive state (i.e. do not need fit)."""
 
-    :params lang: language code.
+    @abc.abstractmethod
+    def transform(self, input: typing.Any, **kwargs: dict):
+        """Abstract base method, need to be implemented in subclass."""
+        return
 
-    :return stop_list: list of stop words.
-    """
-    return many_stop_words.get_stop_words(lang)
+    def __call__(self, input: typing.Any, **kwargs: dict) -> list:
+        """Call function."""
+        return self.transform(input, **kwargs)
 
 
-def tokenizer(raw: str) -> list:
-    """
-    Word tokenization.
+class StatefulProcessUnit(StatelessProcessUnit, metaclass=abc.ABCMeta):
+    """Process unit do persive state (i.e. need fit)."""
 
-    :param raw: raw input given by user.
+    @abc.abstractmethod
+    def fit(self, input: typing.Any) -> dict:
+        """Abstract base method, need to be implemented in subclass."""
+        return
 
-    :return tokens: list of tokenized tokens.
-    """
-    return nltk.word_tokenize(raw)
+    def __call__(self,
+                 input: typing.Any,
+                 **kwargs: dict) -> typing.Union[list, dict]:
+        """Call function."""
+        return self.transform(input, **kwargs), self.fit(input)
```

Comment:
I create `__call__` since it's similar as the previous chain method, I do not necessarily to "call" a specific method later. e.g.
```python
processor = chain(
    ProcessorUnit1(),
    ProcessorUnit2(),
    ProcessorUnit3()
)

rv = processor("This is An example String 123  to be normalized!")
```
the syntax is similar as what we have in `2.0/processors`:
```python
>>> processor = chain(tokenizer,
        ...                   to_lowercase,
        ...                   remove_punctuation,
        ...                   stemming,
        ...                   remove_digits)
        >>> rv = processor("An Example sentence to BE cleaned!")
```

Reply:
Ok I get the point. If that's the motivation, then the problem is the inconsistent signatures of these two `__call__`s. It's not easy to mix them within the same `chain` without some dirty hack on your current implementation of `chain`. Unless you're using chain in a way that either it contains all stateless ones, or all stateful ones, or a chain of any number of stateless ones with a single stateful one at the end. Even that's the real use case, since two types of chains return different things, it's better to make some difference name-wise or usage-wise.

Reply:
actually I already have the first version of chain, still need to refactor:
```python
class chain(object):
    def __init__(self, *process_units):
        self._process_units = []
        self._current_unit_index = 0
        self.context = {}
        for unit in process_units:
            if not isinstance(unit, StatelessProcessUnit):
                raise TypeError("")
            self._process_units.append(unit)
    
    def _next(self):
        """"""
        self._current_unit_index += 1
        return self._process_units[self._current_unit_index]
    
    def _current_unit(self):
        return self._process_units[self._current_unit_index]
    
    def _is_last_unit(self):
        return (self._current_unit_index + 1 >= len(self._process_units))
    
    def _is_stateful_unit(self):
        return isinstance(
            self._process_units[self._current_unit_index + 1], StatefulProcessUnit)
    
    def __call__(self, input):
        rv = self._current_unit()(input)
        while not self._is_last_unit():
            if not self._is_stateful_unit():
                rv = self._next()(rv)
            else:
                rv, ctx = self._next()(rv)
                self.context.update(ctx)
        return rv, self.context
    
    def __len__(self):
        return len(self._process_units)
    
    def append(self, process_unit):
        return self._process_units.append(process_unit)
```

Reply:
I see. It makes sense. If you decides to hide `__call__` from the users and only use it internally, then it's ok. Still some redundancy here, but let's get it working first.